README.md
# http\_relay

Relay HTTP get requests from localhost to a remote host (act as reverse HTTP proxy).

## ROS parameters
- `~host` (mandatory): The remote host to connect to (e.f. "ntrip.pecny.cz").
- `~port` (mandatory): The remote host's port to connect to.
- `~local_port` (default: `~port`): The localhost port to be used for the relay.
- `~num_threads` (default: 8 threads): Number of threads servicing the incoming requests.
- `~buffer_size` (default: 1 byte): Size of the buffer used for reading responses. Generally, a larger buffer should be more efficient, but if it is too large, the local clients may time out before they receive any data.
- `~sigkill_timeout` (default: 0 seconds): If nonzero, the node will be sigkilled in this number of seconds to prevent being stuck on a hanged connection. Use in conjunction with "respawn" roslaunch flag.
- `~sigkill_on_stream_stop` (default: False): If True, `sigkill_timeout` will not be counted when no requests are active, and during requests, each successful data transmission will reset the timeout. This can be used to detect stale streams if you expect an application to be constantly receiving data.

CHANGELOG.md
<!--
SPDX-License-Identifier: BSD-3-Clause
SPDX-FileCopyrightText: Czech Technical University in Prague
-->

# Changelog

## 2.1.3

- Fixed wrongly discovered test classes.

## 2.1.2

- Synchronized maintainer info in setup.py and pyproject.toml .

## 2.1.1

- Fixed CI.

## 2.1.0

- Converted the free functions to class `HttpRelay`.
- Added integration tests.

## 2.0.0

- Converted to pure Python package.

## 1.0.1

- Noetic compatibility.
- Added sigkill\_on\_stream\_stop so that the relay is only killed if there is a stale request.
- Added sigkill\_timeout for automatic restarting of the node even if it hangs on a connection.
- Support NTRIP in Python 3
- Fix Python3 NTRIP relay.
- Initial commit.


test/test_relay.py
# SPDX-License-Identifier: BSD-3-Clause
# SPDX-FileCopyrightText: Czech Technical University in Prague

"""
Integration test testing the relay functionality.
"""

import http_relay
import socket
import threading
import time
import unittest

from http_relay.__main__ import main

try:
    from http.server import HTTPServer, BaseHTTPRequestHandler
    from http.client import *
except ImportError:
    from SimpleHTTPServer import HTTPServer, BaseHTTPRequestHandler
    from httplib import HTTPConnection


class TestHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        self.send_response(200)
        self.end_headers()
        self.wfile.write(b"Test")


class NTRIPHandler(BaseHTTPRequestHandler):
    """This handler simulates NTRIP server responses which use a slight
    modification of the HTTP protocol."""
    def do_GET(self):
        self.wfile.write(b"ICY 200 OK\r\n")
        self.end_headers()
        self.wfile.write(b"Test")


class TestServer(threading.Thread):

    def __init__(self, host, port, handler=TestHandler):
        """
        Create and run the test server thread.
        :param str host: The host to listen on.
        :param int port: The port to listen on.
        """
        threading.Thread.__init__(self)
        HTTPServer.address_family = \
            socket.AF_INET if ":" not in host else socket.AF_INET6
        self.server = HTTPServer((host, port), handler)
        self.daemon = True
        self.running = False
        self.start()

    def run(self):
        """
        Process the server requests.
        """
        try:
            self.running = True
            self.server.serve_forever()
        except Exception:
            pass


class RelayThread(threading.Thread):

    def __init__(self, host, server_port, relay_port, use_main=False):
        """
        Create and run the test relay. All args are passed to `run()`.
        """
        threading.Thread.__init__(self)
        self.host = host
        self.server_port = server_port
        self.relay_port = relay_port
        self.use_main = use_main
        self.daemon = True
        self.running = False
        self.start()

    def run(self):
        """
        Process the server requests.
        """
        self.running = True
        if self.use_main:
            cli_args = list(map(str, [
                self.host, self.server_port, self.relay_port, self.host]))
            main(cli_args)
        else:
            http_relay.run(self.host, self.relay_port, self.host,
                           self.server_port, num_threads=1, buffer_size=1)


class TestRelay(unittest.TestCase):
    def test_local_relay_hostname(self):
        self.do_test('localhost', 8080, 8081)

    def test_local_relay_ipv4(self):
        self.do_test('127.0.0.1', 8040, 8041)

    def test_local_relay_ipv6(self):
        self.do_test('::1', 8060, 8061)

    def test_local_relay_ipv6_brackets(self):
        self.do_test('[::1]', 8050, 8051, "::1")

    def test_local_relay_ntrip(self):
        self.do_test('localhost', 2101, 2102, handler=NTRIPHandler)

    def test_main(self):
        self.do_test('localhost', 8090, 8091, use_main=True)

    def do_test(self, host, server_port, relay_port, server_host=None,
                use_main=False, handler=TestHandler):
        if server_host is None:
            server_host = host
        server_thread = TestServer(server_host, server_port, handler=handler)
        relay_thread = RelayThread(
            host, server_port, relay_port, use_main=use_main)

        while not server_thread.running or not relay_thread.running:
            time.sleep(0.01)
        time.sleep(1.0)

        if not issubclass(handler, TestHandler):
            HTTPConnection.response_class = \
                http_relay.relay.NonstandardHttpResponse
        conn = HTTPConnection(server_host, relay_port, timeout=1.0)
        conn.request("GET", "test")
        resp = conn.getresponse()

        self.assertEqual(200, resp.status)

        resp_body = b""
        while True:
            chunk = resp.read(1)
            if not chunk:
                break
            resp_body += chunk

        self.assertEqual(b"Test", resp_body)


if __name__ == '__main__':
    unittest.main()


src/http_relay/__main__.py
# SPDX-License-Identifier: BSD-3-Clause
# SPDX-FileCopyrightText: Czech Technical University in Prague

import argparse
import logging
import threading

from http_relay import run, sigkill_after


def get_parser():
    parser = argparse.ArgumentParser(
        description="Relay HTTP requests from localhost to a remote host (act as reverse HTTP proxy).")

    parser.add_argument("host", type=str, help='The remote host to connect to (e.g. "cvut.cz")')
    parser.add_argument("port", type=int, default=80, nargs='?',
                        help='The remote host\'s port to connect to (e.g. 80).')
    parser.add_argument("local_port", type=int, default=0, nargs='?',
                        help='The local port to be used for the relay. If left out, will be the same as remote port.')
    parser.add_argument("local_addr", type=str, default='0.0.0.0', nargs='?',
                        help='Local interface (IP or hostname) to run the relay on. By default, it runs on all IPv4 '
                             'interfaces (0.0.0.0).')
    parser.add_argument("-n", "--num-threads", type=int, default=8,
                        help='Number of threads servicing the incoming requests.')
    parser.add_argument("-b", "--buffer-size", type=int, default=1,
                        help='Size of the buffer used for reading responses. Generally, a larger buffer should be more '
                             'efficient, but if it is too large, the local clients may time out before they '
                             'receive any data.')
    parser.add_argument("-t", "--sigkill-timeout", type=int, required=False,
                        help='If specified, the relay will be sigkilled in this number of seconds.')
    parser.add_argument("-s", "--sigkill-on-stream-stop", action="store_true",
                        help='If True, --sigkill-timeout will not be counted when no requests are active, and '
                             'during requests, each successful data transmission will reset the timeout. This can be '
                             'used to detect stale streams if you expect an application to be constantly receiving '
                             'data.')
    return parser


def main(cli_args=None):
    parser = get_parser()
    args = parser.parse_args(cli_args)

    host = args.host.lstrip("[").rstrip("]")  # strip [] from IPv6 addresses
    port = args.port
    local_port = args.local_port if args.local_port != 0 else port
    local_addr = args.local_addr.lstrip("[").rstrip("]")  # strip [] from IPv6 addresses
    num_threads = args.num_threads
    buffer_size = args.buffer_size
    sigkill_timeout = args.sigkill_timeout
    sigkill_on_stream_stop = args.sigkill_on_stream_stop

    logging.info("Relaying HTTP requests from %s:%i to %s:%i using %i threads" % (
        local_addr if ":" not in local_addr else ("[" + local_addr + "]"), local_port,  # wrap IPv6 in []
        host if ":" not in host else ("[" + host + "]"), port,  # wrap IPv6 in []
        num_threads))

    if sigkill_timeout is not None:
        logging.info("HTTP relay has sigkill timeout set to %i seconds. After that time%s, the node will be killed." % (
            sigkill_timeout, " with a stale stream" if sigkill_on_stream_stop else ""))
        t = threading.Thread(target=sigkill_after, args=(sigkill_timeout, sigkill_on_stream_stop))
        t.daemon = True
        t.start()

    run(local_addr, local_port, host, port, num_threads, buffer_size)


if __name__ == '__main__':
    logging.basicConfig(level=logging.INFO)
    main()


src/http_relay/relay.py
# SPDX-License-Identifier: BSD-3-Clause
# SPDX-FileCopyrightText: Czech Technical University in Prague

"""
Relay HTTP get requests from localhost to a remote host (act as reverse HTTP proxy).
"""

import errno
import logging
import os
import signal
import socket
import sys
import threading
import time

try:
    from http.server import HTTPServer, BaseHTTPRequestHandler
    from http.client import *
except ImportError:
    from SimpleHTTPServer import HTTPServer, BaseHTTPRequestHandler
    from httplib import HTTPConnection


__all__ = ['run', 'sigkill_after']


# Some servers (e.g. NTRIP, Shoutcast...) do respond with nonstandard status lines like "ICY 200 OK" instead of
# the standard "HTTP/1.0 200 OK". The Python 3 client raises an exception when it finds these status lines.
# Python 2 client works fine, but overrides the status line to HTTP/1.0. What we want instead is to tell the rest
# of the codebase to look at such response as HTTP/1.0 response, but save the original status line so that it can
# be relayed exactly as it was received.

# Code borrowed and modified from Python 3 source code available under the Python Software Foundation License.
class NonstandardHttpResponse(HTTPResponse):
    def _read_status(self):
        _MAXLINE = 65536
        if sys.version_info.major > 2:
            line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
        else:
            line = self.fp.readline(_MAXLINE + 1)
        if len(line) > _MAXLINE:
            raise LineTooLong("status line")
        if self.debuglevel > 0:
            print("reply:", repr(line))
        if not line:
            # Presumably, the server closed the connection before
            # sending a valid response.
            if sys.version_info.major > 2:
                raise RemoteDisconnected("Remote end closed connection without"
                                         " response")
            else:
                raise BadStatusLine("No status line received - the server "
                                    "has closed the connection")
        self._status_line = line
        try:
            version, status, reason = line.split(None, 2)
        except ValueError:
            try:
                version, status = line.split(None, 1)
                reason = ""
            except ValueError:
                # empty version will cause next test to fail.
                version = ""
        if not version.startswith("HTTP/"):
            # Here is the difference to the original method: we treat invalid version as HTTP 1.0
            version = "HTTP/1.0"

        # The status code is a three-digit number
        try:
            status = int(status)
            if status < 100 or status > 999:
                raise BadStatusLine(line)
        except ValueError:
            raise BadStatusLine(line)
        return version, status, reason


request_num = 0
total_bytes = 0
num_open_requests = 0
shutting_down = False
lock = threading.Lock()


class HTTP10Connection(HTTPConnection):
    _http_vsn_str = "HTTP/1.0"
    _http_vsn = 10
    response_class = NonstandardHttpResponse


class HTTP11Connection(HTTPConnection):
    _http_vsn_str = "HTTP/1.1"
    _http_vsn = 11
    response_class = NonstandardHttpResponse


def is_server_running(server):
    try:
        return server.running
    except:
        return not server._BaseServer__shutdown_request


class Handler(BaseHTTPRequestHandler):
    """
    The main logic of the relay - forward the HTTP request to the remote server with changed Host: header and pass back
    whatever it returns.
    """
    host = "localhost"
    port = 80
    relay_port = 80
    read_buffer_size = 1

    def __init__(self, request, client_address, server):
        global request_num
        self._req_num = request_num
        request_num += 1
        try:
            BaseHTTPRequestHandler.__init__(self, request, client_address, server)
        except socket.error as e:
            self.log_socket_error(e)

    # Do not log requests using the BaseHTTPRequestHandler logging mechanism, we have our own.
    def log_request(self, code='-', size='-'):
        pass

    def log_error(self, format, *args):
        """
        Log an error message.
        :param str format: Format string.
        :param List[Any] args: % parameters of the format string.
        """
        if not shutting_down and is_server_running(self.server):
            logging.error(("Request [%i] error: " + format) % ((self._req_num,) + args))

    def log_socket_error(self, e):
        """
        Log an error raised by socket operation.
        :param socket.error e: The error.
        """
        # Ignore EPIPE and ECONNRESET as that is generated when the other end stops being interested in our data
        if isinstance(e, tuple) and e[0] in (errno.EPIPE, errno.ECONNRESET):
            logging.info("Response [%i]: finished" % (self._req_num,))
        elif ("Errno %i" % (errno.EPIPE,)) in str(e) or ("Errno %i" % (errno.ECONNRESET,)) in str(e):
            logging.info("Response [%i]: finished" % (self._req_num,))
        else:
            if not shutting_down and is_server_running(self.server):
                self.log_error("%s", str(e))

    def log_message(self, format, *args):
        """
        Log an info message.
        :param str format: Format string.
        :param List[Any] args: % parameters of the format string.
        """
        if is_server_running(self.server):
            logging.info(("Request [%i]: " + format) % ((self._req_num,) + args))

    def log_response(self, format, *args):
        """
        Log an info message related to the response.
        :param str format: Format string.
        :param List[Any] args: % parameters of the format string.
        """
        if is_server_running(self.server):
            logging.info(("Response [%i]: " + format) % ((self._req_num,) + args))

    def send_status_line(self, response):
        self.log_request(response.status)
        if self.request_version != 'HTTP/0.9':
            if sys.version_info[0] > 2:
                # self.send_response_only(code, message)
                if not hasattr(self, '_headers_buffer'):
                    self._headers_buffer = []
                self._headers_buffer.append(response._status_line.encode('latin-1', 'strict'))
            else:
                self.wfile.write(response._status_line)

    def do_GET(self):
        """
        Do the relaying work.
        """
        global lock
        global num_open_requests
        with lock:
            num_open_requests += 1
        try:
            # Choose the right HTTP version
            connection_class = HTTP11Connection if self.protocol_version == "HTTP/1.1" else HTTP10Connection
            conn = connection_class(Handler.host, Handler.port)

            # Forward the request with the same headers
            headers = dict(zip(self.headers.keys(), self.headers.values()))

            # Replace host in Host header
            orig_host = None
            host = Handler.host if ":" not in Handler.host else ("[" + Handler.host + "]")
            host_port = host
            for header in headers:
                if header.lower() == "host":
                    orig_host = headers[header]
                    # append port if it is non-default or if it differs from the relay port
                    if Handler.port != 80 and (Handler.port != Handler.relay_port or ":" in orig_host):
                        # : is also valid in IPv6 addresses; a port is specified in an IPv6 only if the last : is after
                        # the last ]
                        if not (":" in orig_host and "]" in orig_host) or (orig_host.rfind(':') > orig_host.rfind(']')):
                            host_port += ":" + str(Handler.port)
                    headers[header] = host_port
                    break

            self.log_message("GET http://%s%s", host_port, self.path)
            conn.request("GET", self.path, headers=headers)

            # Obtain the response
            resp = conn.getresponse()
            self.send_status_line(resp)
            self.log_response("%i %s", resp.status, resp.reason)

            # Forward back the response headers
            for header, value in resp.getheaders():
                # Replace host in Location header
                if orig_host is not None and header.lower() == "location":
                    value = value.replace(host, orig_host, 1)
                self.send_header(header, value)
            self.end_headers()

            # Forward back the response body
            num_bytes = 0
            global total_bytes
            while True:
                chunk = resp.read(Handler.read_buffer_size)
                if not chunk:
                    self.log_response("finished")
                    break
                self.wfile.write(chunk)
                num_bytes += Handler.read_buffer_size
                total_bytes += Handler.read_buffer_size
                if num_bytes > 10 * Handler.read_buffer_size:
                    logging.debug("Response body [%i]: Sent %i bytes." % (self._req_num, num_bytes))
        except socket.error as e:
            self.log_socket_error(e)
        except KeyboardInterrupt:
            pass
        #except Exception as e:
        #    self.log_error("%s", str(e))
        finally:
            with lock:
                num_open_requests -= 1


class Thread(threading.Thread):
    """
    The HTTP server servicing thread.
    """

    def __init__(self, server):
        """
        Create and run the servicing thread.
        :param HTTPServer server: The server to work with.
        """
        threading.Thread.__init__(self)
        self.server = server
        self.daemon = True
        self.start()

    def run(self):
        """
        Process the server requests.
        """
        try:
            self.server.serve_forever()
        except Exception as e:
            if not shutting_down and is_server_running(self.server):
                logging.error("Error in processing thread: " + str(e))


def run(relay_addr, relay_port, remote_host, remote_port, num_threads, buffer_size):
    """
    Run the multithreaded relay server.
    :param str relay_addr: IP address or hostname specifying the local interface(s) to run the relay on
                           (pass 0.0.0.0 or :: to run it on all interfaces (IPv4 or IPv6)).
    :param int relay_port: The local port.
    :param str remote_host: The remote host name.
    :param int remote_port: The remote host port.
    :param int num_threads: Number of servicing threads.
    :param int buffer_size: Size of the buffer used for reading responses. If too large, the forwarding can be too slow.
    """
    server_address = (relay_addr.lstrip("[").rstrip("]"), relay_port)
    Handler.host = remote_host.lstrip("[").rstrip("]")
    Handler.port = remote_port
    Handler.relay_port = relay_port
    Handler.read_buffer_size = buffer_size

    global shutting_down

    try:
        # Create a standalone socket shared by all servers
        socket_type = socket.AF_INET if ":" not in relay_addr else socket.AF_INET6
        sock = socket.socket(socket_type, socket.SOCK_STREAM)
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        sock.bind(server_address)
        sock.listen(5)

        # Create the servers and run their servicing threads
        servers = []
        threads = []
        for i in range(num_threads):
            httpd = HTTPServer(server_address, Handler, False)
            httpd.socket = sock
            httpd.server_bind = httpd.server_close = lambda self: None
            servers.append(httpd)
            threads.append(Thread(httpd))

        # Wait for node exit
        try:
            while not shutting_down:
                time.sleep(0.1)
        except KeyboardInterrupt:
            pass

        logging.info("Stopping HTTP relay.")

        shutting_down = True

        # First, shut down the socket, which should convince server.shutdown() to finish.
        sock.shutdown(socket.SHUT_RDWR)
        sock.close()

        # Shut down the servers (their service threads are daemons, so we don't need to join them)
        for server in servers:
            if server is not None:
                server.shutdown()

    except socket.gaierror as e:
        logging.error(str(e))
        shutting_down = True
        sys.exit(2)
    except socket.error as e:
        logging.error(str(e))
        shutting_down = True
        sys.exit(1)
    except:
        shutting_down = True
        raise


def shutdown():
    global shutting_down
    shutting_down = True


def sigkill_after(timeout, check_streaming=False):
    global total_bytes
    global num_open_requests
    global shutting_down
    remaining = timeout
    last_total_bytes = -1
    while not shutting_down:
        if not check_streaming or (num_open_requests > 0 and total_bytes == last_total_bytes):
            remaining -= 1
            if remaining <= 0:
                logging.error("Stopping HTTP relay!")
                shutting_down = True
                time.sleep(0.01)
                os.kill(os.getpid(), signal.SIGKILL)
                return
        else:
            remaining = timeout
        if check_streaming and remaining == timeout // 2:
            logging.warning("Relayed HTTP stream stopped. The relay will be stopped in %i sec if the stream does not "
                            "resume." % (timeout // 2,))
        last_total_bytes = total_bytes
        time.sleep(1)
